6746 Client Code Checks

GeronimoJohn opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

Fixes #6746 

# 👩🏻‍💻 What does this PR do?

We initially set up a Typescript compilation and linter hook (`pre-commit-lint`) that runs every time a developer makes a change within the client directory and then commit their changes. It's a very powerful tool but the TS compilation and linter takes a bit of time to run.

To avoid waiting. We've decided to create a Github Action that runs a linter and formatter against the client directory. To ensure that we're still following the code standards that was initially set. And remove husky or the `pre-commit-lint` hook from running.

![Screenshot 2025-03-05 at 2 20 59 PM](https://github.com/user-attachments/assets/e66fa96a-cc20-445a-89a3-6fdaecf58cb7)

<!-- Explain the changes you made -->

<!-- why are the changes needed -->

<!-- Add a screenshot if there are UI changes  -->

## 💌 Any notes for the reviewer?

<!-- Do you have any specific questions for the reviewer? -->

<!-- Is there a high risk/complicated change they should focus on? -->

<!-- any general areas of the codebase touched? any side effects caused? -->

<!-- Anything half cooked but going to be finished off in a different PR? -->

ESLint had a major update about a few months ago - version 9.0.0 onwards. It's a breaking change and we are required to migrate our old configuration ESLint file to follow their new pattern. Not doing so will not make ESLint work. The interim solution is to downgrade to version 8.50.0. And are to work on the migration at a different time whenever time permits

# 🧪 Testing

<!-- Explain the steps you'd take to test the changes of this PR manually -->

DEVS ONLY

- [ ] This shouldn't run the pre-build-lint when doing a commit after a change within the client directory
- [ ] There should be a new Github Action called - `Client Code Checks`
   - [ ] It should pass if it doesn't see any changes that doesn't follow the rules set in our ESLint config
      - [ ] Update any file within the client directory and commit
      - [ ] Remove changes after 
   - [ ] It should fail if it does see any changes that doesn't follow the rules set in our ESLint config
   - [ ] Ensure that the logs returned by the Github action makes sense

# 📃 Documentation

- [ ] **Part of an epic**: documentation will be completed for the feature as a whole
- [ ] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [ ] **These areas should be updated or checked**: <!-- _(e.g.)_ New `issued` column in `Requisitions` indicates stock quantity already in shipments -->
  1.
  2.


# 📃 Reviewer Checklist

The PR Reviewer(s) should fill out this section before approving the PR

**Breaking Changes**
- [ ] No Breaking Changes in the Graphql API
- [ ] Technically some Breaking Changes but not expected to impact any integrations

**Issue Review**
- [ ] All requirements in original issue have been covered
- [ ] A follow up issue(s) have been created to cover additional requirements

**Tests Pass**
- [ ] Postgres
- [ ] SQLite
- [ ] Frontend

GeronimoJohn commented:
**Note on Logs:**

You'll notice a lot of `warnings` in the logs. These are not ideal but currently acceptable. The majority of these warnings come from `useEffect` hooks, specifically their dependency arrays. `useEffect` and ESLint rules can be tricky, as dependency arrays often depend on the specific use case.  

If a `useEffect` is working as expected without the dependencies ESLint suggests, it's fine to ignore the warning using TypeScript's `ts-ignore`, e.g.  
`// eslint-disable-next-line react-hooks/exhaustive-deps`.

jmbrunskill commented on client/.husky/pre-commit:
```diff
@@ -1,9 +1,13 @@
 . "$(dirname "$0")/_/husky.sh"
 
-SRC_PATTERN="client/"
+# The following lines have been commented out to disable automatic execution of the pre-commit linting and formatting tasks.
```
Can we just remove husky now? Is it needed still?

GeronimoJohn replied:
From what I can see, not needed anymore. I was going to suggest that we just remove it now

GeronimoJohn replied:
Have removed it. Please see changes here: https://github.com/msupply-foundation/open-msupply/pull/6796/commits/5969e2980af3f26e056f8004ef22771b4ec6ae7b

I've dropped these libraries
- Husky
- Lint-staged

CarlosNZ commented on client/packages/inventory/src/Stocktake/DetailView/modal/StocktakeLineEdit/StocktakeLineEditTables.tsx:
```diff
@@ -26,7 +26,7 @@ import {
   InventoryAdjustmentReasonSearchInput,
   ItemVariantInputCell,
-  PackSizeEntryCell,
+  packSizeEntryCell,
```
This really should be `PackSizeEntryCell`, as the source is very much a component. I wonder if we just need to correctly type the original component in "PackSizeEntryCell.tsx" as React.FC etc...

GeronimoJohn replied:
That makes sense actually

GeronimoJohn replied:
This proved to be a bit trickier than expected.

Here are a few things I tried:

1. Since `PackSizeEntryCell` is a HOC, the initial return is a function.
   - Making it start with a capital letter causes ESLint to give the error when we call it: `A function with a name starting with an uppercase letter should only be used as a constructor. eslint(new-cap)`
      - The solution is to disable the error with `// eslint-disable-next-line new-cap`
   - Additionally, since the actual component is in a callback, converting the first character to an uppercase gives another error: hooks cannot be called inside a callback.
   
2. Converting it to just be a normal React function component creates a type mismatch, as shown in the image below, which might require more work:
   - ![Screenshot 2025-03-06 at 11 50 44 AM](https://github.com/user-attachments/assets/95b336e1-3927-41a4-91b5-ecd96461ed31)

The solution I applied:
- Keep the first character of the component in uppercase.
- Disable the error with `// eslint-disable-next-line new-cap`.
- Convert the function called inside to a standalone component, which we then call.
  - If we stick to lowercase for the first character, we adhere to the rules of React for hooks in callbacks, which is weird to me lol.

GeronimoJohn replied:
This proved to be a bit trickier than expected.

Here are a few things I tried:

1. Since `PackSizeEntryCell` is a HOC, the initial return is a function.
   - Making it start with a capital letter causes ESLint to give the error when we call it: `A function with a name starting with an uppercase letter should only be used as a constructor. eslint(new-cap)`
      - The solution is to disable the error with `// eslint-disable-next-line new-cap`
   - Additionally, since the actual component is in a callback, converting the first character to an uppercase gives another error: hooks cannot be called inside a callback.
   
2. Converting it to just be a normal React function component creates a type mismatch, as shown in the image below, which might require more work:
   - ![Screenshot 2025-03-06 at 11 50 44 AM](https://github.com/user-attachments/assets/95b336e1-3927-41a4-91b5-ecd96461ed31)

The solution I applied:
- Keep the first character of the component in uppercase.
- Disable the error with `// eslint-disable-next-line new-cap`.
- Convert the function called inside to a standalone component, which we then call.
  - If we stick to lowercase for the first character, we adhere to the rules of React for hooks in callbacks, which is weird to me lol.

CarlosNZ commented on client/packages/requisitions/src/RequestRequisition/DetailView/IndicatorEdit/IndicatorLineEdit.tsx:
```diff
@@ -123,14 +124,18 @@ export const IndicatorLineEdit = ({
-            autoFocus={i === 0}
-          />
-        ))}
+        {columns.map((column, i) => {
+          return column.value != null ? (
+            <InputWithLabel
+              key={column.value?.id}
+              data={column}
+              disabled={disabled}
+              autoFocus={i === 0}
+            />
+          ) : (
+            <></>
+          );
+        })}
```
```suggestion
        {columns.map(
          (column, i) =>
            column.value != null && (
              <InputWithLabel
                key={column.value?.id}
                data={column}
                disabled={disabled}
                autoFocus={i === 0}
              />
            )
        )}
```
Bit cleaner?

CarlosNZ commented on client/packages/requisitions/src/RnRForms/api/hooks/useRnRFormContext.ts:
```diff
@@ -18,6 +18,7 @@ export const useRnRFormContext = create<RnRFormContext>(set => ({
     })),
   clearDraftLine: id =>
     set(state => {
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
This is a bit odd, because we have a lint rule to ignore unused errors that are "_":
```
    '@typescript-eslint/no-unused-vars': [
      'error',
      {
        argsIgnorePattern: '^_',
        caughtErrorsIgnorePattern: '^_',
      },
    ],
```
Maybe have a quick look what's going on there? But probably don't spend too much time on it, this is fine otherwise.

CarlosNZ reviewed (commented):
Looking good, @GeronimoJohn, this is going to be a big help. Few little comments.

I think we should probably wait to merge this until after 2.6.0 has been merged into develop. (And then we can do all the re-formatting as well :)

GeronimoJohn commented on .github/workflows/client-code-checks.yaml:
```diff
@@ -0,0 +1,30 @@
+on:
+  pull_request:
+    branches:
+      - main
+      - develop
```
Ah that's a good point, I'll look into that

GeronimoJohn commented on client/packages/requisitions/src/RnRForms/api/hooks/useRnRFormContext.ts:
```diff
@@ -18,7 +18,6 @@ export const useRnRFormContext = create<RnRFormContext>(set => ({
   clearDraftLine: id =>
     set(state => {
-      // eslint-disable-next-line @typescript-eslint/no-unused-vars
       const { [id]: _, ...draftLines } = state.draftLines;
```
Updated the rule for it to support this case

```typescript
'@typescript-eslint/no-unused-vars': [
      'error',
      {
        argsIgnorePattern: '^_',
        caughtErrorsIgnorePattern: '^_',
        varsIgnorePattern: '^_',
      },
    ]
```

GeronimoJohn commented:
Thanks for the review Carl, please see changes here: https://github.com/msupply-foundation/open-msupply/pull/6796/commits/229e61ad545ba4b13a52129f21401e0273c38b55

GeronimoJohn commented on .github/workflows/client-code-checks.yaml:
```diff
@@ -0,0 +1,32 @@
+      - main
+      - develop
+      # This pattern matches branches for release candidates (e.g., v2.6.0-RC)
+      - v*-RC
```
Supports cases for RC branches. This workflow will run against any PRs that are open and targeting these branches.

As of the time this was written, the GitHub Action is failing because v2.6.0-RC has a console log.

andreievg replied:
Why not on every PR ?

GeronimoJohn replied:
Thats a good question actually, it should just be applied to every PR. Will change soon

GeronimoJohn commented on .github/workflows/server-tests.yaml:
```diff
@@ -3,8 +3,10 @@ on:
     branches:
       - main
       - develop
+      # This pattern matches branches for release candidates (e.g., v2.6.0-RC)
+      - v*-RC
```
Side Quest: Enable CI or server tests to run against RC branches

andreievg commented on client/.eslintrc.js:
```diff
@@ -40,6 +40,7 @@ module.exports = {
       {
         argsIgnorePattern: '^_',
         caughtErrorsIgnorePattern: '^_',
+        varsIgnorePattern: '^_',
```
What does this do ?

andreievg replied:
Never mind oop i should read all comments found above

andreievg commented:
@GeronimoJohn, in this PR: https://github.com/msupply-foundation/open-msupply/pull/6879, I've added re-compile, which i think should just replace the compile yarn script (it's 3 times as fast), I think it's worth doing that here, what you reckon ?

GeronimoJohn commented:
> @GeronimoJohn, in this PR: #6879, I've added re-compile, which i think should just replace the compile yarn script (it's 3 times as fast), I think it's worth doing that here, what you reckon ?

Sounds good to me! Also noticed in that PR that we're formatting the graphql generated files 🔥. Big fan

GeronimoJohn commented on .github/workflows/server-tests.yaml:
```diff
@@ -1,10 +1,8 @@
-      - main
-      - develop
+    types: [opened, reopened, edited]
     paths:
-      - 'server/**'
+      - "server/**"
```
Side Quest: As discussed on omSupply Devs, we want server tests to run on every PR. Added opened, reopened, and edited types to this action. This GH action runs whenever a PR affecting the server directory is opened, reopened, or edited.

GeronimoJohn commented on .github/workflows/server-tests.yaml:
```diff
@@ -1,10 +1,8 @@
-      - develop
+    types: [opened, reopened, edited, synchronize]
     paths:
-      - 'server/**'
+      - "server/**"
```
Side Quest: Run this on every PR that is opened, reopened, edited, or synchronized. This ensures the action runs on all PR changes.

andreievg commented:
@GeronimoJohn after updating from develop could you please replace current 'compile' yarn script with 're-compile'

GeronimoJohn commented:
Thanks @andreievg for the review, 

Updated Github action `Client Code Checks` and `Server Tests` to run on every PR: https://github.com/msupply-foundation/open-msupply/pull/6796/commits/f2a86a36dd9d488c7b64bc4b64461522ee001536

Have also updated the yarn command for `lint-and-format` to use `re-compile`.

Noticed that some of our ColdChain queries are not following camelCase. Added a fix here: https://github.com/msupply-foundation/open-msupply/pull/6796/commits/23bb7ba41122e4c1159b9ab35fe46f0dc1682411

CarlosNZ commented on client/package.json:
```diff
@@ -16,8 +16,6 @@
     "compile": "lerna run --scope @openmsupply-client/* --parallel tsc --since HEAD -- --noEmit",
     "re-compile": "tsc --project packages/host/tsconfig.json --noEmit",
     "compile-full": "lerna run --scope @openmsupply-client/* --parallel tsc",
```
Just wondering what we want "re-compile" for, as opposed to "compile-full"?

andreievg replied:
We don't want any parallel compilation it's 3 x slower

CarlosNZ commented on client/packages/invoices/src/InboundShipment/DetailView/modals/InboundLineEdit/TabTables.tsx:
```diff
@@ -118,7 +113,7 @@ export const QuantityTableComponent: FC<TableProps> = ({
     ],
     getColumnLookupWithOverrides('packSize', {
-      Cell: PackUnitEntryCell,
+      Cell: PackSizeEntryCell<DraftInboundLine>,
```
Great work getting this mess sorted! 👍

CarlosNZ reviewed (approved):
Looks good @GeronimoJohn, will be good to get this in. I'm happy with the changes, don't suppose @andreievg has anything remaining?
